server/src/main/java/org/elasticsearch/index/fielddata/plain/SortedNumericDVIndexFieldData.java

```diff
@@ -134,11 +135,63 @@ public AtomicNumericFieldData load(LeafReaderContext context) {
                 return new SortedNumericFloatFieldData(reader, field);
             case DOUBLE:
                 return new SortedNumericDoubleFieldData(reader, field);
+            case DATE_NANOSECONDS:
+                return new NanoSecondFieldData(reader, field, numericType);
             default:
                 return new SortedNumericLongFieldData(reader, field, numericType);
         }
     }
 
+    /**
+     * A small helper class that can be configured to load nanosecond field data either in nanosecond resolution retaining the original
+     * values or in millisecond resolution converting the nanosecond values to milliseconds
+     */
+    public final class NanoSecondFieldData extends AtomicLongFieldData {
+
+        private final LeafReader reader;
+        private final String fieldName;
+
+        NanoSecondFieldData(LeafReader reader, String fieldName, NumericType numericType) {
+            super(0L, numericType);
+            this.reader = reader;
+            this.fieldName = fieldName;
+        }
+
+        @Override
+        public SortedNumericDocValues getLongValues() {
+            try {
```

Comment:
nit: it's easier to read if you reuse the `getLongValuesAsNanos()`?:
```
final SortedNumericDocValues dv = getLongValuesAsNanos();
return new AbstractSortedNumericDocValues()...
````

Reply:
done.